### PR TITLE
Improve split operator by oneDNN reorder primitive

### PR DESCRIPTION
**Pull request by bgawrych**
## Description ##
Split operation can be done by multiple reorder primitives with properly set strides

image

Benchmark script:
```
import gc
import mxnet
import mxnet.gluon.nn as nn
import mxnet.numpy as np
import time


dims = [128, 512, 1024]
print("shape;axis;time")
for ndim in range(2):
   for dim1 in dims:
     for dim2 in dims:
        shape = (dim1, dim2) if ndim == 0 else (8, dim1, dim2)
        a = np.random.uniform(-1.0, 1.0, shape).astype(np.float32)
        for axis in range(2 + ndim):
            for section in (2, 4, 8):
                tic = time.time()
                for i in range(100):
                    out = np.array_split(a, axis=axis, indices_or_sections=section)
                    [o.wait_to_read() for o in out]
                toc = time.time()
                gc.collect()
                print(f"{shape};{axis};{section};{toc-tic}")

```

### CODE REVIEW

**Comment by bgawrych** [comment]
@mxnet-bot run ci [centos-gpu]

**Review by agrabows — APPROVED** [review]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-gpu, centos-gpu]

File: src/operator/nn/dnnl/dnnl_split.cc [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  const auto fwd = DNNLSplitFwd::GetCached(param, tensors, split_pts, split_axis);
+  fwd.Execute(tensors, split_pts, split_axis, req);
+}
+
+DNNLSplitFwd::Tensors::Tensors(const NDArray& input, const std::vector<NDArray>& outputs)
+    : input(input), outputs(outputs) {}
+
+typedef ParamOpSign<SplitParam> DNNLSplitSignature;
+
+DNNLSplitFwd DNNLSplitFwd::GetCached(const SplitParam& param,
```
Comment (PawelGlomski-Intel) [inline src/operator/nn/dnnl/dnnl_split.cc:61]:
```suggestion
DNNLSplitFwd& DNNLSplitFwd::GetCached(const SplitParam& param,
```

**Review by PawelGlomski-Intel — COMMENTED** [review]
Missing reference?

File: src/operator/nn/dnnl/dnnl_split.cc [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+                           const int split_axis,
+                           const std::vector<OpReqType>& req) const {
+  const auto& cpu_engine = CpuEngine::Get()->get_engine();
+
+  const auto& input_tensor = tensors.input.Reorder2Default();
+  int out_idx = 0, primitive_idx = 0;
+  int axis_offset      = strides[split_axis] * GetTypeSize(input_tensor.dtype());
+  std::byte* input_ptr = reinterpret_cast<std::byte*>(input_tensor.data().dptr_);
+
+  for (const auto& out : tensors.outputs) {
```
Comment (PawelGlomski-Intel) [inline src/operator/nn/dnnl/dnnl_split.cc:127]: Could you summarize in a sentence or two what happens in this loop? How exactly is the reorder primitive used here?

Comment (bgawrych) [reply src/operator/nn/dnnl/dnnl_split.cc:127]: Split op can split data into multiple different tensors basing on passed axis and split points. Split points can be irregular e.g. we can split tensor of shape [4, 6, 6] into tensors [4, 1, 6], [4, 2, 6] and [4,3,6]. We need different reorder primitives to have different dimensions of output tensors and as there is reorder2default I'm passing address with offset to dnnl memory to copy memory from input to output

File: src/operator/nn/dnnl/dnnl_split.cc [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  const auto cpu_engine = CpuEngine::Get()->get_engine();
+  const auto input      = tensors.input.Reorder2Default();
+  const auto& ishape    = input.shape();
+  const auto& dtype     = get_dnnl_type(input.dtype());
+  const auto format_tag = static_cast<dnnl::memory::format_tag>(GetDefaultFormat(ishape.ndim()));
+
+  strides = dnnl::memory::dims(ishape.ndim(), 1);
+  // last dim stride = 1, start loop from the penultimate
+  for (int i = ishape.ndim() - 2; i >= 0; --i) {
+    strides[i] = strides[i + 1] * ishape[i + 1];
```
Comment (bartekkuncer) [inline src/operator/nn/dnnl/dnnl_split.cc:94]: Starting `i` at `ishape.ndim() - 1` would trade two additions for one subtraction.

File: src/operator/tensor/matrix_op.cc [hunk 4]
```diff
@@ -1177,6 +1178,32 @@ Example::
...
+  }
+}
+
+inline static bool SplitInferStorageType(const nnvm::NodeAttrs& attrs,
+                                         const int dev_mask,
+                                         DispatchMode* dispatch_mode,
+                                         std::vector<int>* in_attrs,
+                                         std::vector<int>* out_attrs) {
+  return DNNLStorageType(
+      attrs, dev_mask, /*support onednn*/ true, dispatch_mode, in_attrs, out_attrs);
```
Comment (bartekkuncer) [inline src/operator/tensor/matrix_op.cc:1203]: oneDNN

File: src/operator/nn/dnnl/dnnl_split.cc (hunk 3 above)
Comment (PawelGlomski-Intel) [reply src/operator/nn/dnnl/dnnl_split.cc:94]: I think the current version is more intuitive.

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-gpu, windows-gpu]

**Review by PawelGlomski-Intel — APPROVED** [review]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [windows-gpu]

**Review by bartekkuncer — APPROVED** [review]
LGTM. Please change `onednn` to `oneDNN` in any future change.
